File: src/lib/aws/services/cost-explorer.ts
import {
  GetCostAndUsageCommand,
  GetCostForecastCommand,
  GetDimensionValuesCommand,
  type GetCostAndUsageCommandInput,
  type GetCostForecastCommandInput,
  type DateInterval,
  type Granularity,
  type Metric
} from '@aws-sdk/client-cost-explorer';
import { getCostExplorerClient } from '../clients';
import { format, subDays, startOfMonth, endOfMonth } from 'date-fns';

export interface CostData {
  date: string;
  amount: number;
  unit: string;
  service?: string;
}

export interface ServiceCost {
  service: string;
  amount: number;
  percentage: number;
}

export interface CostSummary {
  total: number;
  byService: ServiceCost[];
  trend: CostData[];
  forecast?: number;
  currency: string;
}

// Get daily cost data for a date range
export const getDailyCosts = async (
  startDate: Date,
  endDate: Date,
  groupBy?: 'SERVICE' | 'LINKED_ACCOUNT' | 'REGION'
): Promise<CostData[]> => {
  const client = getCostExplorerClient();

  const input: GetCostAndUsageCommandInput = {
    TimePeriod: {
      Start: format(startDate, 'yyyy-MM-dd'),
      End: format(endDate, 'yyyy-MM-dd')
    },
    Granularity: 'DAILY',
    Metrics: ['UnblendedCost'],
    GroupBy: groupBy ? [{ Type: 'DIMENSION', Key: groupBy }] : undefined
  };

  try {
    const command = new GetCostAndUsageCommand(input);
    const response = await client.send(command);

    const costData: CostData[] = [];

    if (response.ResultsByTime) {
      response.ResultsByTime.forEach(result => {
        const date = result.TimePeriod?.Start || '';

        if (result.Groups && result.Groups.length > 0) {
          // Grouped data
          result.Groups.forEach(group => {
            const amount = parseFloat(group.Metrics?.UnblendedCost?.Amount || '0');
            const unit = group.Metrics?.UnblendedCost?.Unit || 'USD';
            const service = group.Keys?.[0] || 'Unknown';

            costData.push({
              date,
              amount,
              unit,
              service
            });
          });
        } else {
          // Non-grouped data
          const amount = parseFloat(result.Total?.UnblendedCost?.Amount || '0');
          const unit = result.Total?.UnblendedCost?.Unit || 'USD';

          costData.push({
            date,
            amount,
            unit
          });
        }
      });
    }

    return costData;
  } catch (error) {
    console.error('Error fetching cost data:', error);
    return [];
  }
};

// Get cost breakdown by service for current month
export const getServiceCosts = async (): Promise<ServiceCost[]> => {
  const client = getCostExplorerClient();

  const now = new Date();
  const startDate = startOfMonth(now);
  const endDate = endOfMonth(now);

  const input: GetCostAndUsageCommandInput = {
    TimePeriod: {
      Start: format(startDate, 'yyyy-MM-dd'),
      End: format(endDate, 'yyyy-MM-dd')
    },
    Granularity: 'MONTHLY',
    Metrics: ['UnblendedCost'],
    GroupBy: [{ Type: 'DIMENSION', Key: 'SERVICE' }],
    Filter: {
      Not: {
        Dimensions: {
          Key: 'RECORD_TYPE',
          Values: ['Credit', 'Refund']
        }
      }
    }
  };

  try {
    const command = new GetCostAndUsageCommand(input);
    const response = await client.send(command);

    const serviceCosts: ServiceCost[] = [];
    let totalCost = 0;

    if (response.ResultsByTime && response.ResultsByTime[0]) {
      const result = response.ResultsByTime[0];

      if (result.Groups) {
        // Calculate total cost
        result.Groups.forEach(group => {
          const amount = parseFloat(group.Metrics?.UnblendedCost?.Amount || '0');
          totalCost += amount;
        });

        // Calculate service costs with percentages
        result.Groups.forEach(group => {
          const service = group.Keys?.[0] || 'Unknown';
          const amount = parseFloat(group.Metrics?.UnblendedCost?.Amount || '0');
          const percentage = totalCost > 0 ? (amount / totalCost) * 100 : 0;

          if (amount > 0.01) { // Filter out negligible costs
            serviceCosts.push({
              service,
              amount,
              percentage
            });
          }
        });

        // Sort by amount descending
        serviceCosts.sort((a, b) => b.amount - a.amount);
      }
    }

    return serviceCosts;
  } catch (error) {
    console.error('Error fetching service costs:', error);
    return [];
  }
};

// Get cost forecast for next month
export const getCostForecast = async (): Promise<number> => {
  const client = getCostExplorerClient();

  const now = new Date();
  const startDate = new Date(now.getFullYear(), now.getMonth() + 1, 1); // First day of next month
  const endDate = new Date(now.getFullYear(), now.getMonth() + 2, 0); // Last day of next month

  const input: GetCostForecastCommandInput = {
    TimePeriod: {
      Start: format(startDate, 'yyyy-MM-dd'),
      End: format(endDate, 'yyyy-MM-dd')
    },
    Metric: 'UNBLENDED_COST',
    Granularity: 'MONTHLY'
  };

  try {
    const command = new GetCostForecastCommand(input);
    const response = await client.send(command);

    if (response.Total?.Amount) {
      return parseFloat(response.Total.Amount);
    }

    return 0;
  } catch (error) {
    console.error('Error fetching cost forecast:', error);
    return 0;
  }
};

// Get comprehensive cost summary
export const getCostSummary = async (
  timeRange: '7d' | '30d' | 'mtd' | '3m' = '30d'
): Promise<CostSummary> => {
  let startDate: Date;
  const endDate = new Date();

  switch (timeRange) {
    case '7d':
      startDate = subDays(endDate, 7);
      break;
    case '30d':
      startDate = subDays(endDate, 30);
      break;
    case 'mtd': // Month to date
      startDate = startOfMonth(endDate);
      break;
    case '3m':
      startDate = subDays(endDate, 90);
      break;
  }

  // Get daily costs
  const dailyCosts = await getDailyCosts(startDate, endDate);

  // Get service breakdown
  const serviceCosts = await getServiceCosts();

  // Get forecast
  const forecast = await getCostForecast();

  // Calculate total
  const total = dailyCosts.reduce((sum, cost) => sum + cost.amount, 0);

  return {
    total,
    byService: serviceCosts,
    trend: dailyCosts,
    forecast,
    currency: 'USD'
  };
};

// Get specific service cost details
export const getServiceCostDetails = async (
  serviceName: string,
  timeRange: '7d' | '30d' | '3m' = '30d'
): Promise<CostData[]> => {
  const client = getCostExplorerClient();

  let startDate: Date;
  const endDate = new Date();

  switch (timeRange) {
    case '7d':
      startDate = subDays(endDate, 7);
      break;
    case '30d':
      startDate = subDays(endDate, 30);
      break;
    case '3m':
      startDate = subDays(endDate, 90);
      break;
  }

  const input: GetCostAndUsageCommandInput = {
    TimePeriod: {
      Start: format(startDate, 'yyyy-MM-dd'),
      End: format(endDate, 'yyyy-MM-dd')
    },
    Granularity: timeRange === '7d' ? 'DAILY' : 'MONTHLY',
    Metrics: ['UnblendedCost', 'UsageQuantity'],
    Filter: {
      Dimensions: {
        Key: 'SERVICE',
        Values: [serviceName]
      }
    },
    GroupBy: [{ Type: 'DIMENSION', Key: 'USAGE_TYPE' }]
  };

  try {
    const command = new GetCostAndUsageCommand(input);
    const response = await client.send(command);

    const costData: CostData[] = [];

    if (response.ResultsByTime) {
      response.ResultsByTime.forEach(result => {
        const date = result.TimePeriod?.Start || '';

        if (result.Groups) {
          result.Groups.forEach(group => {
            const amount = parseFloat(group.Metrics?.UnblendedCost?.Amount || '0');
            const unit = group.Metrics?.UnblendedCost?.Unit || 'USD';
            const service = group.Keys?.[0] || serviceName;

            if (amount > 0) {
              costData.push({
                date,
                amount,
                unit,
                service
              });
            }
          });
        }
      });
    }

    return costData;
  } catch (error) {
    console.error(`Error fetching cost details for ${serviceName}:`, error);
    return [];
  }
};

// Get cost anomalies (significant cost increases)
export const detectCostAnomalies = async (
  threshold: number = 0.2 // 20% increase threshold
): Promise<{ date: string; service: string; increase: number }[]> => {
  const endDate = new Date();
  const startDate = subDays(endDate, 14); // Get 2 weeks of data

  const costs = await getDailyCosts(startDate, endDate, 'SERVICE');

  const anomalies: { date: string; service: string; increase: number }[] = [];

  // Group costs by service
  const serviceMap = new Map<string, CostData[]>();
  costs.forEach(cost => {
    if (cost.service) {
      if (!serviceMap.has(cost.service)) {
        serviceMap.set(cost.service, []);
      }
      serviceMap.get(cost.service)?.push(cost);
    }
  });

  // Detect anomalies for each service
  serviceMap.forEach((serviceCosts, service) => {
    serviceCosts.sort((a, b) => a.date.localeCompare(b.date));

    for (let i = 1; i < serviceCosts.length; i++) {
      const current = serviceCosts[i];
      const previous = serviceCosts[i - 1];

      if (previous.amount > 0) {
        const increase = (current.amount - previous.amount) / previous.amount;

        if (increase > threshold) {
          anomalies.push({
            date: current.date,
            service,
            increase: increase * 100 // Convert to percentage
          });
        }
      }
    }
  });

  return anomalies;
};

File: packages/backend/internal/aws/costexplorer.go
package aws

import (
	"context"
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/costexplorer"
	"github.com/aws/aws-sdk-go-v2/service/costexplorer/types"
)

// CostExplorerClient wraps the Cost Explorer client
type CostExplorerClient struct {
	client *costexplorer.Client
}

// NewCostExplorerClient creates a new Cost Explorer client
func NewCostExplorerClient(cfg aws.Config) *CostExplorerClient {
	return &CostExplorerClient{
		client: costexplorer.NewFromConfig(cfg),
	}
}

// CostData represents AWS cost information
type CostData struct {
	TotalCost      float64                `json:"totalCost"`
	Currency       string                 `json:"currency"`
	Services       []ServiceCost          `json:"services"`
	DailyCosts     []DailyCost            `json:"dailyCosts"`
	Period         string                 `json:"period"`
}

// ServiceCost represents cost breakdown by service
type ServiceCost struct {
	ServiceName string  `json:"serviceName"`
	Cost        float64 `json:"cost"`
	Percentage  float64 `json:"percentage"`
}

// DailyCost represents daily cost data
type DailyCost struct {
	Date string  `json:"date"`
	Cost float64 `json:"cost"`
}

// GetCostAndUsage retrieves cost and usage data
func (c *CostExplorerClient) GetCostAndUsage(ctx context.Context, startDate, endDate time.Time) (*CostData, error) {
	// Format dates for AWS API
	start := startDate.Format("2006-01-02")
	end := endDate.Format("2006-01-02")

	costData := &CostData{
		Currency: "USD",
		Period:   fmt.Sprintf("%s to %s", start, end),
	}

	// Get total cost and daily breakdown
	dailyInput := &costexplorer.GetCostAndUsageInput{
		TimePeriod: &types.DateInterval{
			Start: &start,
			End:   &end,
		},
		Granularity: types.GranularityDaily,
		Metrics:     []string{"UnblendedCost"},
	}

	dailyResult, err := c.client.GetCostAndUsage(ctx, dailyInput)
	if err != nil {
		return nil, fmt.Errorf("failed to get daily costs: %w", err)
	}

	// Process daily costs
	var totalCost float64
	for _, result := range dailyResult.ResultsByTime {
		if result.TimePeriod != nil && result.Total != nil {
			if costAmount, ok := result.Total["UnblendedCost"]; ok {
				if costAmount.Amount != nil {
					cost := parseFloat(*costAmount.Amount)
					dailyCost := DailyCost{
						Date: *result.TimePeriod.Start,
						Cost: cost,
					}
					costData.DailyCosts = append(costData.DailyCosts, dailyCost)
					totalCost += cost
				}
			}
		}
	}
	costData.TotalCost = totalCost

	// Get cost breakdown by service
	serviceInput := &costexplorer.GetCostAndUsageInput{
		TimePeriod: &types.DateInterval{
			Start: &start,
			End:   &end,
		},
		Granularity: types.GranularityMonthly,
		Metrics:     []string{"UnblendedCost"},
		GroupBy: []types.GroupDefinition{
			{
				Type: types.GroupDefinitionTypeTag,
				Key:  aws.String("SERVICE"),
			},
		},
	}

	serviceResult, err := c.client.GetCostAndUsage(ctx, serviceInput)
	if err != nil {
		// Log error but continue with available data
		fmt.Printf("Failed to get service breakdown: %v\n", err)
	} else {
		// Process service costs
		for _, result := range serviceResult.ResultsByTime {
			for _, group := range result.Groups {
				if group.Metrics != nil {
					if costAmount, ok := group.Metrics["UnblendedCost"]; ok {
						if costAmount.Amount != nil && len(group.Keys) > 0 {
							cost := parseFloat(*costAmount.Amount)
							serviceCost := ServiceCost{
								ServiceName: group.Keys[0],
								Cost:        cost,
								Percentage:  (cost / totalCost) * 100,
							}
							costData.Services = append(costData.Services, serviceCost)
						}
					}
				}
			}
		}
	}

	return costData, nil
}

// GetForecast retrieves cost forecast data
func (c *CostExplorerClient) GetForecast(ctx context.Context, days int) (*CostData, error) {
	// Calculate date range
	startDate := time.Now().AddDate(0, 0, 1) // Start from tomorrow
	endDate := startDate.AddDate(0, 0, days-1)

	start := startDate.Format("2006-01-02")
	end := endDate.Format("2006-01-02")

	input := &costexplorer.GetCostForecastInput{
		TimePeriod: &types.DateInterval{
			Start: &start,
			End:   &end,
		},
		Metric:      types.MetricUnblendedCost,
		Granularity: types.GranularityDaily,
	}

	result, err := c.client.GetCostForecast(ctx, input)
	if err != nil {
		return nil, fmt.Errorf("failed to get cost forecast: %w", err)
	}

	costData := &CostData{
		Currency: "USD",
		Period:   fmt.Sprintf("%s to %s (forecast)", start, end),
	}

	// Process forecast data
	if result.Total != nil && result.Total.Amount != nil {
		costData.TotalCost = parseFloat(*result.Total.Amount)
	}

	for _, forecast := range result.ForecastResultsByTime {
		if forecast.TimePeriod != nil && forecast.MeanValue != nil {
			dailyCost := DailyCost{
				Date: *forecast.TimePeriod.Start,
				Cost: parseFloat(*forecast.MeanValue),
			}
			costData.DailyCosts = append(costData.DailyCosts, dailyCost)
		}
	}

	return costData, nil
}

// GetServiceCosts retrieves costs for specific services
func (c *CostExplorerClient) GetServiceCosts(ctx context.Context, services []string, startDate, endDate time.Time) ([]ServiceCost, error) {
	start := startDate.Format("2006-01-02")
	end := endDate.Format("2006-01-02")

	var serviceCosts []ServiceCost

	for _, service := range services {
		input := &costexplorer.GetCostAndUsageInput{
			TimePeriod: &types.DateInterval{
				Start: &start,
				End:   &end,
			},
			Granularity: types.GranularityMonthly,
			Metrics:     []string{"UnblendedCost"},
			Filter: &types.Expression{
				Dimensions: &types.DimensionValues{
					Key:    types.DimensionService,
					Values: []string{service},
				},
			},
		}

		result, err := c.client.GetCostAndUsage(ctx, input)
		if err != nil {
			fmt.Printf("Failed to get cost for service %s: %v\n", service, err)
			continue
		}

		var totalCost float64
		for _, r := range result.ResultsByTime {
			if r.Total != nil {
				if costAmount, ok := r.Total["UnblendedCost"]; ok {
					if costAmount.Amount != nil {
						totalCost += parseFloat(*costAmount.Amount)
					}
				}
			}
		}

		serviceCosts = append(serviceCosts, ServiceCost{
			ServiceName: service,
			Cost:        totalCost,
		})
	}

	// Calculate percentages
	var total float64
	for _, sc := range serviceCosts {
		total += sc.Cost
	}
	for i := range serviceCosts {
		if total > 0 {
			serviceCosts[i].Percentage = (serviceCosts[i].Cost / total) * 100
		}
	}

	return serviceCosts, nil
}

// parseFloat converts string to float64
func parseFloat(s string) float64 {
	var f float64
	fmt.Sscanf(s, "%f", &f)
	return f
}